bolts/data/datasets/vision/celeba.py

```diff
@@ -21,68 +24,158 @@
+        superclass: CelebAttr = CelebAttr.Smiling,
+        subclass: CelebAttr = CelebAttr.Male,
+        transform: Optional[ImageTform] = None,
+        split: Optional[CelebASplit] = None,
     ) -> None:
-        self.base = Path(root) / CELEBA_BASE_FOLDER
-        super().__init__(root=str(self.base), transform=transform)
+        super().__init__(root=root, transform=transform)
 
+        self.base = Path(root) / self._BASE_FOLDER
+        self._img_dir = self.base / "img_align_celeba"
         self.superclass = superclass
         self.subclass = subclass
 
-        dataset, self._img_dir = em.celeba(
-            download_dir=root,
-            label=superclass,  # type: ignore
-            sens_attr=subclass,  # type: ignore
-            download=False,  # we'll download manually
-            check_integrity=False,  # we'll check manually
-        )
-        assert dataset is not None, "could not load CelebA"
-
         if download:
             self._download_and_unzip_data()
         elif not self._check_unzipped():
             raise RuntimeError(
                 f"Data don't exist at location {self.base.resolve()}. " "Have you downloaded it?"
             )
 
-        # load meta data
-        data_tup = dataset.load(labels_as_features=True)
-        self.metadata = data_tup.x
-
-        self.x = self.metadata["filename"].to_numpy(copy=True)
-        self.s = torch.as_tensor(data_tup.s.to_numpy(), dtype=torch.int32).view(-1)
-        self.y = torch.as_tensor(data_tup.y.to_numpy(), dtype=torch.int32).view(-1)
-
+        if split is None:
+            mask = slice(None)
+        else:
+            # splits: information about which samples belong to train, val or test
+            splits = (
+                pd.read_csv(
+                    self.base / "list_eval_partition.txt", delim_whitespace=True, names=["split"]
+                )
+                .to_numpy()
+                .squeeze()
+            )
+            mask = splits == split.value
+        attrs = pd.read_csv(
+            self.base / "list_attr_celeba.txt",
+            delim_whitespace=True,
+            header=1,
+            usecols=[superclass.name, subclass.name],
+        ).iloc[
+            mask
+        ]  # type: ignore
+
+        self.x = np.array(attrs.index)
+        self.s = torch.tensor(attrs[subclass.name])
+        self.s = torch.tensor(attrs[superclass.name])
+        # map from {-1, 1} to {0, 1}
+        self.s = torch.div(self.s + 1, 2, rounding_mode='floor')
+        self.y = torch.div(self.s + 1, 2, rounding_mode='floor')
+
+        # infer the appropriate image-loading backend based on the type of 'transform'
         self._il_backend: ImageLoadingBackend = infer_il_backend(self.transform)
 
     def _check_unzipped(self) -> bool:
         return self._img_dir.is_dir()
 
     def _download_and_unzip_data(self) -> None:
         """Attempt to download data if files cannot be found in the base directory."""
+        if self._check_unzipped():
```

Comment:
I purposefully didn't include that check because I wanted cached_download to do its md5 check.

Reply:
Yes, I did; I got annoyed with the message about the MD5 hash being checked appearing every time the code was run with download=True.

Reply:
👍